### DISCUSSION TITLE
Adding python3 pip, snap packages to an Ubuntu 22.04 image

### DISCUSSION THREAD

**Discussion by chmjlg** [systemd/mkosi#3114] [Q&A] [answered]
# mkosi commit the issue has been seen with
https://github.com/systemd/mkosi/commit/029cf06632f6ae79c69f444c84fb36bcce893a0e

# Used host distribution
Ubuntu 24.04

# Used target distribution
Ubuntu 22.04

# Linux kernel version used
6.8.40-generic

# Question
I'd like to have a few snap packages installed in my Ubuntu 22.04 image. I attempted using a mkosi.prepare and mkosi.postinst script to list out the snap packages I'd like to download (yq, and helm) but was unable to download them using snap. I wonder if these scripts are the correct locations to have the packages installed in the generated image. 

# Used mkosi config
```
# this is the main configuration file.

[Distribution]
Distribution=ubuntu
Release=jammy
Repositories=main,universe

[Output]
Format=disk
OutputDirectory=./output

[Content]
Bootable=yes
Hostname=temp
WithNetwork=yes
EnvironmentFiles=./install.env
RootPassword=123

#WithRecommends=true
#Bootloader=grub
Packages=
  apt
  bash
  ca-certificates
  curl
  dbus
  less
  linux-base
  linux-sound-base
  linux-image-6.8.0-40-generic
  linux-headers-6.8.0-40-generic
  linux-modules-extra-6.8.0-40-generic
  linux-modules-6.8.0-40-generic
  linux-tools-6.8.0-40-generic
  linux-tools-common
  util-linux
  nano
  ssh
  sudo
  systemd
  udev
  wget
  network-manager
  net-tools
  iproute2
  iputils-ping
  wireless-tools
  wpasupplicant
  ethtool
  avahi-daemon
  avahi-utils
  avahi-autoipd
  netplan.io
  isc-dhcp-client
  isc-dhcp-common
  dnsutils
  libpcap0.8
  tcpd
  vim
  pciutils
  pci.ids
  linux-firmware
  rfkill
  rsync
  gparted
  parted
  build-essential
  bash-completion
  git
  bluez
  bluez-tools
  joystick
  usbutils
  lshw
  fdisk
  grub-efi
  cloud-guest-utils
  gcc-12
  python3-pip
  dosfstools
  apt-utils
  e2fsprogs
  python3-paho-mqtt
  snapd
  software-properties-common

[Partitions]
RootSize=5G

[Build]
ToolsTree=default
ToolsTreeDistribution=ubuntu
ToolsTreePackages=python3-pip,apt-transport-https,ca-certificates,curl,gnupg,lsb-release,software-properties-common
```

**Comment by chmjlg** [comment]
My apologies I gave an outdated mkosi.conf file attached are the updated mkosi.conf, postinst script, and a debug file that has the output of mkosi --force --debug. I initially wanted to try using snap inside the image, so I use the chroot extension. Realizing I didn't have network acces, i removed the chroot extension and tried running in the host. 
[debug.txt](https://github.com/user-attachments/files/17297738/debug.txt)
# mkosi.conf
```
# this is the main configuration file.

[Distribution]
Distribution=ubuntu
Release=jammy
Repositories=main,universe

[Output]
Format=disk
OutputDirectory=./output

[Content]
Bootable=yes
Hostname=talos
WithNetwork=yes
EnvironmentFiles=./install.env
RootPassword=123

#WithRecommends=true
#Bootloader=grub
Packages=
  apt
  bash
  ca-certificates
  curl
  dbus
  less
  linux-base
  linux-sound-base
  linux-image-6.8.0-40-generic
  linux-headers-6.8.0-40-generic
  linux-modules-extra-6.8.0-40-generic
  linux-modules-6.8.0-40-generic
  linux-tools-6.8.0-40-generic
  linux-tools-common
  util-linux
  nano
  ssh
  sudo
  systemd
  udev
  wget
  network-manager
  net-tools
  iproute2
  iputils-ping
  wireless-tools
  wpasupplicant
  ethtool
  avahi-daemon
  avahi-utils
  avahi-autoipd
  netplan.io
  isc-dhcp-client
  isc-dhcp-common
  dnsutils
  libpcap0.8
  tcpd
  vim
  pciutils
  pci.ids
  linux-firmware
  rfkill
  rsync
  gparted
  parted
  build-essential
  bash-completion
  git
  can-utils
  bluez
  bluez-tools
  joystick
  usbutils
  lshw
  fdisk
  grub-efi
  cloud-guest-utils
  gcc-12
  python3-pip
  dosfstools
  apt-utils
  e2fsprogs
  python3-paho-mqtt
  snapd

[Partitions]
RootSize=5G

[Build]
ToolsTree=default
ToolsTreeDistribution=ubuntu
ToolsTreePackages=python3-pip,apt-transport-https,ca-certificates,curl,gnupg,lsb-release
```
# mkosi.postinst
```
#!/bin/bash

snap install helm --classic
snap install yq
```

**Reply by behrmann** [reply]
mkosi's scripts act on the image from the outside, so if snap has a `--root` switch to point it to the tree it should work on, that's what you would need to use. If it doesn't have it or you don't want to use it for some other reason, you would need a `mkosi.postinst.chroot` script or use `mkosi-chroot` in the postinst script.

**Reply by chmjlg** [reply]
Got it, ideally I'd like to list all snap packages in a script and keep everything contained within the image. I'll go with the postinst script.

**Comment by chmjlg** [comment]
File of debug text without the chroot extension 
[debugwithoutchroot.txt](https://github.com/user-attachments/files/17297922/debugwithoutchroot.txt)

**Comment by daandemeyer** [comment, marked as answer]
@chmjlg Use `WithNetwork=yes` to enable network access

**Reply by chmjlg** [reply]
Gotcha thanks @DaanDeMeyer , I did enable it on my mkosi.conf under Hostname in the Content section, but I don't get network access in the script if I chroot into the image. If I don't chroot, I do get network access since I'm running out of the host. Might be an issue in my VM setup since I get DNS resolution failure.
